posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,372 @@
+
+Data.table provides the fwrite() function for writing data to a file while Pandas offers the to_csv() function for writing data to a CSV file.
+
+## Comparison code
+
+```{r,warning=FALSE,message=FALSE}
+write.colors <- c(
+  "data.table::fwrite" = "#D6604D",
+  "pandas::to_csv" = "#BF812D"
+)
+file_path = 'data.csv'
+n.rows <- 100
+seconds.limit <- 10
+atime.write.vary.cols <- atime::atime(
+  N = as.integer(10^seq(2, 10, by = 0.2)),
+  setup = {
+    set.seed(1)
+    input.vec <- rnorm(n.rows * N)
+    input.mat <- matrix(input.vec, n.rows, N)
+    input.df <- data.frame(input.mat) 
+    pd <- import("pandas")
+    input_df_pd <- r_to_py(input.df)
+  },
+  seconds.limit = seconds.limit,
+  "data.table::fwrite" = {
+    data.table::fwrite(input.df, tempfile(), showProgress = FALSE)
+  },
+  "pandas::to_csv" = {
+    input_df_pd$to_csv(file_path, index = FALSE)
+  }
+)
+```
+
+```{r,warning=FALSE,message=FALSE}
+refs.write.vary.cols <- atime::references_best(atime.write.vary.cols)
+plot(refs.write.vary.cols)
+```
+
+atime::references_best() : This function returns a list, which can then be visualized using plot().
+The plot shows the asymptotic reference lines, which defines the estimated asymptotic time and memory complexity. The output and figure will includes violet reference lines, in which the text labels can be interpreted in terms of big O notation (asymptotic time and memory usage).
+
+The graph compares the performance of data.table::fwrite and pandas::to_csv for writing data. The top left graph shows that pandas::to_csv has a sharp increase in time as the dataset size increases, while data.table::fwrite remains relatively constant. The top right graph shows both methods with almost constant times, but pandas::to_csv uses more kilobytes. Overall, data.table::fwrite performs significantly faster than pandas::to_csv, particularly for larger datasets, suggesting that it may be a more efficient method for writing data.
+
+
+The predict() function is used to generate predictions based on the atime::reference_best() dataset. The resulting plot illustrates the data size, N, that can be processed within a specific time or memory limit.
+
+
+```{r,warning=FALSE,message=FALSE}
+pred.write.vary.cols <- predict(refs.write.vary.cols)
+plot(pred.write.vary.cols)
+```
+
+```{r}
+gg.write.dt.pd <- plot(pred.write.vary.cols) +
+  theme(text = element_text(size = 15)) +
+  ggtitle(sprintf("Write real numbers to CSV, with pandas in Python \nand data.table in R, %d x N", n.rows)) +
+  scale_x_log10("N = number of columns to write") +
+  scale_y_log10("Computation time (seconds)\nmedian line, min/max band\nover 10 timings") +
+  facet_null() +
+  scale_fill_manual(values = write.colors) +
+  scale_color_manual(values = write.colors)
+
+print(gg.write.dt.pd)
+```
+
+The plot above shows that in terms of writing data, the data.table package in R outperforms Python's pandas library and particularly useful when dealing with large datasets.
```

Comment:
I would rather say 'the pandas library in Python' since Pandas is not a part of Python (just like data.table is not a part of base R, and you mentioned 'the data.table package in R')

Reply:
yes, thanks